creusot-contracts/src/ghost_ptr.rs

```diff
@@ -0,0 +1,171 @@
+    #[requires(self@.contains(ptr))]
+    #[ensures(*result == *self@.lookup_unsized(ptr))]
+    pub fn ptr_as_ref(&self, ptr: *const T) -> &T {
+        unsafe { &*ptr }
+    }
+
+    /// Shrinks the view of the `self` so that it's model is now new-model
+    #[trusted]
+    #[requires(new_model.subset(self@))]
+    #[ensures(result@ == *new_model)]
+    pub fn shrink_token_ref(&self, new_model: Ghost<FMap<*const T, T>>) -> &GhostPtrToken<T> {
+        self
+    }
+
+    /// Mutably borrows `ptr` and shrinks `t` so that it can no longer be used to access `ptr`
+    // Safety no other token has permission to `self`
+    // `t` can no longer be used to access `ptr`
+    #[trusted]
+    #[requires((**self)@.contains(ptr))]
+    #[ensures(*result == *(**self)@.lookup_unsized(ptr))]
+    #[ensures((*^self)@ == (**self)@.remove(ptr))]
+    #[ensures((^*self)@ == (^^self)@.insert(ptr, ^result))]
+    // #[ensures(!(^^t)@.contains(ptr))]
+    // ^~ It shouldn't have been possible to add pointer to `t` while we were holding a mutable reference to the pointer
+    pub fn take_mut<'o, 'i>(self: &'o mut &'i mut GhostPtrToken<T>, ptr: *const T) -> &'i mut T {
+        unsafe { &mut *(ptr as *mut _) }
+    }
+
+    /// Mutably borrows `ptr`
+    #[requires(self@.contains(ptr))]
+    #[ensures(*result == *(*self)@.lookup_unsized(ptr))]
+    #[ensures((^self)@ == (*self)@.insert(ptr, ^result))]
+    pub fn ptr_as_mut(&mut self, ptr: *const T) -> &mut T {
+        let mut t = self;
+        t.take_mut(ptr)
+    }
+
+    /// Transfers ownership of `ptr` back into a `Box`
+    // Safety `ptr` is now dangling but since `self` doesn't have permission anymore no token does so this is okay
+    #[trusted]
+    #[requires((*self)@.contains(ptr))]
+    #[ensures(*result == *(*self)@.lookup_unsized(ptr))]
+    #[ensures((^self)@ == (*self)@.remove(ptr))]
+    pub fn ptr_to_box(&mut self, ptr: *const T) -> Box<T> {
+        unsafe { Box::from_raw(ptr as *mut _) }
+    }
+
+    #[trusted]
+    #[ensures((*self)@.disjoint(other@))]
+    // Since we had full permission to and all of the entries in `self` and `other` simultaneously,
+    // no pointer could have been in both
+    #[ensures((^self)@ == (*self)@.union(other@))]
+    pub fn merge(&mut self, other: GhostPtrToken<T>) {}
+
+    /// Leaks memory iff the precondition fails
+    #[requires(self@.is_empty())]
+    pub fn drop(self) {}
+}
+
+impl<T: ?Sized> GhostPtrExt<T> for GhostPtr<T> {
+    #[trusted]
+    #[logic]
+    #[ensures(forall<t: GhostPtrToken<T>> !t@.contains(result))]
+    // #[ensures(result.addr_logic() == 0@)]
+    #[ensures(forall<ptr: GhostPtr<T>> ptr.addr_logic() == result.addr_logic() ==> ptr == result)]
+    fn null_logic() -> Self {
+        absurd
+    }
+
+    #[trusted]
+    #[logic]
+    fn addr_logic(self) -> Int {
+        absurd
+    }
+}
+
+pub trait GhostPtrExt<T: ?Sized>: Sized {
+    #[logic]
+    fn null_logic() -> Self;
+    #[logic]
+    fn addr_logic(self) -> Int;
+}
```

Comment:
Why do you use an extension trait here? Why not directly implement the methods on `GhostPtr`?

Reply:
Right now `GhostPtr<T>` is just a type alias for `* const T`

Reply:
Of course, I forgot.